### PR TITLE
Pin importlib-metadata to avoid an unsolvable dependency resolution by pip

### LINKED ISSUE #4042: New release of opentelemetry-python break the dependency tree.
**Version**
All versions greater than 3.26

**Describe the bug**
The new opentelemetry-api package depends on `importlib-metadata ~= 6.0` where the previous version(1.18) depends on the more restrictive ` ~=6.0.0`. With this, `pip` got confused and installs `importlib-metadata==6.8.0` with some other packages being "incompatible" with this release and requiring `==6.0.1`.

**To Reproduce**
Try to install all pulpcore dependencies from the `main` branch.
ex. `podman run -it python:3.11 /bin/bash -c 'git clone https://github.com/pulp/pulpcore /app ; pip install pip-tools; pip install /app'`

**Expected behavior**
It was expected a clean installation.

**Additional context**
The problem itself starts to be described here --> https://matrix.to/#/!aVApiNMtnstWbwDcVU:matrix.org/$HiAzZwFxytADU4ABBQguZ0hs8JslBZYBSBWwbZySGyY?via=libera.chat&via=matrix.org&via=fedora.im

Also, the upstream issue https://github.com/open-telemetry/opentelemetry-python/issues/3382

### PR DESCRIPTION
**Pull request by decko**
Fixes #4042

### CODE REVIEW

**Review by mdellweg — COMMENTED** [review]
I have questions. How can releasing a new version of the telemetry stack break us if we don't even cover it in the upper bound requirement? That is exactly the kind of problem we set out to avoid using strictly bounded requirements.
I get that the update of opentelemetry-api introduces an unsatisfiable dependency. Can we avoid that using a constraints file?
I wish you could declare incompatibilities on python packages after releasing them.

https://pip.pypa.io/en/stable/user_guide/#constraints-files

**Comment by mdellweg** [comment]
I don't think we can backport upping the lower bound of a dependency.

**Comment by ggainey** [comment]
> I have questions. How can releasing a new version of the telemetry stack break us if we don't even cover it in the upper bound requirement? That is exactly the kind of problem we set out to avoid using strictly bounded requirements. I get that the update of opentelemetry-api introduces an unsatisfiable dependency. Can we avoid that using a constraints file? I wish you could declare incompatibilities on python packages after releasing them.
> 
> https://pip.pypa.io/en/stable/user_guide/#constraints-files

See the issue I opened against otel : https://github.com/open-telemetry/opentelemetry-python/issues/3382 for details.

The net is, otel-0.39b is NOT INSTALLABLE right now. Doesn't matter that we pinned it, it's broken.

**Comment by decko** [comment]
> I have questions. How can releasing a new version of the telemetry stack break us if we don't even cover it in the upper bound requirement? That is exactly the kind of problem we set out to avoid using strictly bounded requirements. I get that the update of opentelemetry-api introduces an unsatisfiable dependency. Can we avoid that using a constraints file? I wish you could declare incompatibilities on python packages after releasing them.
> 
> https://pip.pypa.io/en/stable/user_guide/#constraints-files

I think that `pip` is unable to solve the dependency. `poetry` was able to easily solve it downgrading `importlib-metadata` to 6.0.1. It's not a problem of the opentelemetry package itself, IMHO.

About the constraints file, thanks for the suggestion. I'm trying it right now, but it seems incompatible with the way we're using requirements file and setuptools. Maybe I'm missing something here too, but the documentation about it is too scarce. 

Another solution to this is pinning down the importlib-metadata to 6.0.1.

**Comment by decko** [comment]
> > I don't think we can backport upping the lower bound of a dependency.
> 
> Right now, it is not possible to install pulpcore with otel-0.38b or otel-0.39b. We can change the pins, or we can remove otel, or we can just not ever be able to install/release fixes for previous pulpcore versions.

@ggainey what if we just pin `importlib-metadata` to 6.0.1? Not the best solution, but a quick quirky to allow the CI and releases keep running until we reach a better solution/agreement.

**Comment by ggainey** [comment]
> @ggainey what if we just pin `importlib-metadata` to 6.0.1? Not the best solution, but a quick quirky to allow the CI and releases keep running until we reach a better solution/agreement.

"quick and least-invasive" and "fixes existing releases" are the only things I care about right this second. We can talk about doing something more-correct once we've made our users whole.

**Review by ggainey — APPROVED** [review]
Looks like the least-invasive way to address this.

**Review by mdellweg — APPROVED** [review]
I agree. Can you add a comment to the importlb-metadata that this is a pin for a third party dep, so not to be removed/relaxed without knowing there's an installable solution?

File: CHANGES/4042.bugfix [hunk 1]
```diff
@@ -0,0 +1 @@
+Pin importlib-metadata to 6.0.1 given pip is unable to solve a opentelemetry dependency version correctly.
```
Comment (lubosmj) [inline CHANGES/4042.bugfix:1]:
```suggestion
Pinned importlib-metadata to 6.0.1 given pip is unable to solve an opentelemetry dependency version correctly.
```

Comment (decko) [reply CHANGES/4042.bugfix:1]: Thanks for the observation @lubosmj

**Review by mdellweg — APPROVED** [review]

**Pull request merged**